src/hll.c

```diff
@@ -92,6 +118,182 @@ enum {
+	List *argumentList = NIL;
+	const bool findVariadics = false;
+	const bool findDefaults = false;
+	const bool missingOK = true;
+
+	functionList = FuncnameGetCandidates(qualifiedFunctionNameList, argumentCount,
+										 argumentList, findVariadics,
+										 findDefaults, missingOK);
+
+	if (functionList == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_FUNCTION),
+						errmsg("function \"%s\" does not exist", functionName)));
+	}
+	else if (functionList->next != NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_AMBIGUOUS_FUNCTION),
+						errmsg("more than one function named \"%s\"", functionName)));
+	}
+
+	/* get function oid from function list's head */
+	functionOid = functionList->oid;
+	return functionOid;
+}
+
+/*
+ * MaximizeCostOfHashAggregateWithHLL maximizes the cost of the path if it tries
+ * to run hll aggregate function with hash aggregate.
+ */
+static void
+MaximizeCostOfHashAggregateWithHLL(Path *path)
+{
+	List *varList = pull_var_clause((Node*) path->pathtarget->exprs,
+									PVC_INCLUDE_AGGREGATES);
+	ListCell *varCell = NULL;
+
+	foreach(varCell, varList)
+	{
+		Var *var = (Var *) lfirst(varCell);
+
+		if (nodeTag(var) == T_Aggref)
+		{
+			Aggref *aggref = (Aggref *) var;
+
+			// That part will be cached
+			int argCount = list_length(aggref->args);
+			Oid hllId = get_extension_oid(HLL_EXTENSION_NAME, false);
+			Oid hllSchemaOid = get_extension_schema(hllId);
+			const char *hllSchemaName = get_namespace_name(hllSchemaOid);
+			Oid addFunctionId = FunctionOid(hllSchemaName, HLL_ADD_AGG_NAME, argCount);
+			Oid unionFunctionId = FunctionOid(hllSchemaName, HLL_UNION_AGG_NAME, 1);
+
+			if(aggref->aggfnoid == addFunctionId || aggref->aggfnoid == unionFunctionId)
+			{
+				path->total_cost = INT_MAX;
+			}
+		}
+	}
+}
+
+/*
+ * multi_aggregation_restriction_hook is assigned to create_upper_paths_hook to
+ * check whether there exist a path with hash aggregate. If that aggregate is
+ * introduced by hll, it's cost is maximized to force planner to not to select
+ * hash aggregate.
+ */
+void
+multi_aggregation_restriction_hook(PlannerInfo *root, UpperRelationKind stage,
+								   RelOptInfo *input_rel, RelOptInfo *output_rel)
+{
+	if (DisableHashagg)
+	{
+		if (stage == UPPERREL_GROUP_AGG)
+		{
+			ListCell *pathCell = list_head(output_rel->pathlist);
+			foreach(pathCell, output_rel->pathlist)
+			{
+				Path *path = (Path *) lfirst(pathCell);
+				if(path->pathtype == T_Agg && ((AggPath *)path)->aggstrategy == AGG_HASHED)
+				{
+					MaximizeCostOfHashAggregateWithHLL(path);
+				}
+			}
+		}
+	}
+}
+
+/* Register HLL configuration variables. */
+static void
+RegisterHLLConfigVariables(void)
+{
+	DefineCustomBoolVariable(
+		"hll.disable_hashagg",
+		gettext_noop("Disables using hash aggreagate with hll aggregate functions"),
+		NULL,
+		&DisableHashagg,
+		false,
+		PGC_USERSET,
+		0,
+		NULL, NULL, NULL);
+}
+
+/* _PG_init is the shared library initialization function */
+void
+_PG_init(void)
+{
+	/*
```

Comment:
Will we need to add hll to shared_preload_libraries after this change? If so, we should error out if it is not added to shared_preload_libraries, like [this](https://github.com/citusdata/citus/blob/435a0363285a02be786dddb9863df86f84c48b8d/src/backend/distributed/shared_library_init.c#L149)